### PR TITLE
[fiber] Add support for ARM64 targets

### LINKED ISSUE #1111: Fibers not implemented on ARM64
This breaks fibers support on Raspberry Pi and macOS running on Apple Silicon. This is also a problem for running the unit test suite on these targets.

See https://github.com/modm-io/modm/blob/develop/src/modm/processing/fiber/module.lb#L29.

### PR DESCRIPTION
**Pull request by salkinium**
I've added a new key to the hosted device files: `arch` with `arm64` and `x86_64`. Maybe it should be `aarch64`?

- [x] New targets with `-arm64` suffix, `-x86_64` is implicit for backwards compatibility:
  - [x] `hosted-linux-arm64`
  - [x] `hosted-darwin-arm64`
- [x] Removed `hosted-rpi` target, as wiringPi is deprecated and thus the target lost its only peripheral.
- [x] Tested on Apple Silicon M2
- [x] Test on ARM64 Linux specifically RPi

### CODE REVIEW

**Comment by salkinium** [comment]
I'll merge this tonight, we can check and fix the RPi ARM64 later. It's not our primary target anyways, but I'd like to have the unit tests pass on my machine again ;-P

**Comment by rleh** [comment]
> * [ ]  Test on ARM64 Linux specifically RPi

It does not work 😢, because the x86_64 context switching (`fiber/context_x86_64.cpp`) is choosen for both the `hosted-linux` **and** `hosted-linux-arm64` targets:

<details>

```
user@raspberrypi:~/modm $ (cd test && make run-hosted-linux)
/home/user/.local/bin/lbuild -p ../build/generated-unittest/hosted -c config/hosted.xml -D":target=hosted-linux" -C ../build/generated-unittest/hosted build --no-log
>> modm: Recomputing device cache...

scons -j8 -C ../build/generated-unittest/hosted run
scons: Entering directory `/home/user/modm/build/generated-unittest/hosted'
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
Unittest······· unittest_runner.cpp
╭────Jinja2──── modm/modm_tools/info.c.in
╰───Template──> modm/src/info_git.c
╭────Jinja2──── modm/modm_tools/info.c.in
╰───Template──> modm/src/info_build.c
Compiling C++·· scons-release/modm/src/modm/architecture/driver/atomic/flag.o
Compiling C++·· scons-release/modm/src/modm/architecture/interface/can.o
Compiling C++·· scons-release/modm/src/modm/communication/sab/interface.o
Compiling C++·· scons-release/modm/src/modm/communication/sab/slave.o
Compiling C++·· scons-release/modm/src/modm/communication/xpcc/backend/can/connector.o
Compiling C···· scons-release/modm/src/info_git.o
Compiling C···· scons-release/modm/src/info_build.o
Compiling C++·· scons-release/unittest_runner.o
Compiling C++·· scons-release/modm/src/modm/communication/xpcc/backend/header.o
Compiling C++·· scons-release/modm/src/modm/communication/xpcc/communicator.o
Compiling C++·· scons-release/modm/src/modm/communication/xpcc/dispatcher.o
Compiling C++·· scons-release/modm/src/modm/communication/xpcc/postman/dynamic_postman/dynamic_postman.o
Compiling C++·· scons-release/modm/src/modm/communication/xpcc/postman/postman.o
Compiling C++·· scons-release/modm/src/modm/container/smart_pointer.o
Compiling C++·· scons-release/modm/src/modm/debug/logger/hosted/default_style.o
Compiling C++·· scons-release/modm/src/modm/driver/can/can_lawicel_formatter.o
Compiling C++·· scons-release/modm/src/modm/driver/can/mcp2515.o
Compiling C++·· scons-release/modm/src/modm/driver/io/terminal.o
Compiling C++·· scons-release/modm/src/modm/driver/motor/drv832x_spi.o
Compiling C++·· scons-release/modm/src/modm/driver/temperature/ltc2984.o
Compiling C++·· scons-release/modm/src/modm/io/iostream.o
Compiling C++·· scons-release/modm/src/modm/io/iostream_printf.o
Compiling C++·· scons-release/modm/src/modm/math/geometry/angle.o
Compiling C++·· scons-release/modm/src/modm/math/geometry/vector2.o
Compiling C++·· scons-release/modm/src/modm/math/utils/bit_operation.o
Compiling C++·· scons-release/modm/src/modm/math/utils/pc/operator.o
Compiling C++·· scons-release/modm/src/modm/platform/core/assert.o
Compiling C++·· scons-release/modm/src/modm/platform/core/clock.o
Compiling C++·· scons-release/modm/src/modm/platform/core/memory.o
Compiling C++·· scons-release/modm/src/modm/processing/fiber/context_x86_64.o
modm/src/modm/processing/fiber/context_x86_64.cpp:51:20: warning: 'naked' attribute directive ignored [-Wattributes]
   51 | modm_context_entry()
      |                    ^
modm/src/modm/processing/fiber/context_x86_64.cpp:152:51: warning: 'naked' attribute directive ignored [-Wattributes]
  152 | modm_context_jump(modm_context_t*, modm_context_t*)
      |                                                   ^
/tmp/ccwrMUes.s: Assembler messages:
/tmp/ccwrMUes.s:16: Error: operand 1 must be an integer register -- `mov (%rsp),%rdi'
/tmp/ccwrMUes.s:17: Error: operand 1 must be an integer register -- `mov 8(%rsp),%rsi'
/tmp/ccwrMUes.s:18: Error: unknown mnemonic `jmp' -- `jmp *%rsi'
/tmp/ccwrMUes.s:89: Error: unknown mnemonic `stmxcsr' -- `stmxcsr 0xa0(x1)'
/tmp/ccwrMUes.s:90: Error: unknown mnemonic `fnstcw' -- `fnstcw 0xa4(x1)'
/tmp/ccwrMUes.s:278: Error: unknown mnemonic `leaq' -- `leaq -0xe8(%rsp),%rsp'
/tmp/ccwrMUes.s:279: Error: unknown mnemonic `stmxcsr' -- `stmxcsr 0xa0(%rsp)'
/tmp/ccwrMUes.s:280: Error: unknown mnemonic `fnstcw' -- `fnstcw 0xa4(%rsp)'
/tmp/ccwrMUes.s:281: Error: unknown mnemonic `movq' -- `movq %r12,0xa8(%rsp)'
/tmp/ccwrMUes.s:282: Error: unknown mnemonic `movq' -- `movq %r13,0xb0(%rsp)'
/tmp/ccwrMUes.s:283: Error: unknown mnemonic `movq' -- `movq %r14,0xb8(%rsp)'
/tmp/ccwrMUes.s:284: Error: unknown mnemonic `movq' -- `movq %r15,0xc0(%rsp)'
/tmp/ccwrMUes.s:285: Error: unknown mnemonic `movq' -- `movq %rbx,0xd8(%rsp)'
/tmp/ccwrMUes.s:286: Error: unknown mnemonic `movq' -- `movq %rbp,0xe0(%rsp)'
/tmp/ccwrMUes.s:287: Error: unknown mnemonic `movq' -- `movq %rsp,(%rdi)'
/tmp/ccwrMUes.s:288: Error: unknown mnemonic `movq' -- `movq (%rsi),%rsp'
/tmp/ccwrMUes.s:289: Error: unknown mnemonic `ldmxcsr' -- `ldmxcsr 0xa0(%rsp)'
/tmp/ccwrMUes.s:290: Error: unknown mnemonic `fldcw' -- `fldcw 0xa4(%rsp)'
/tmp/ccwrMUes.s:291: Error: unknown mnemonic `movq' -- `movq 0xa8(%rsp),%r12'
/tmp/ccwrMUes.s:292: Error: unknown mnemonic `movq' -- `movq 0xb0(%rsp),%r13'
/tmp/ccwrMUes.s:293: Error: unknown mnemonic `movq' -- `movq 0xb8(%rsp),%r14'
/tmp/ccwrMUes.s:294: Error: unknown mnemonic `movq' -- `movq 0xc0(%rsp),%r15'
/tmp/ccwrMUes.s:295: Error: unknown mnemonic `movq' -- `movq 0xd8(%rsp),%rbx'
/tmp/ccwrMUes.s:296: Error: unknown mnemonic `movq' -- `movq 0xe0(%rsp),%rbp'
/tmp/ccwrMUes.s:297: Error: unknown mnemonic `leaq' -- `leaq 0xe8(%rsp),%rsp'
scons: *** [scons-release/modm/src/modm/processing/fiber/context_x86_64.o] Error 1
scons: building terminated because of errors.
make: *** [Makefile:29: run-hosted-linux] Error 2
```

</details>

I'll investigate further later this day...

**Comment by rleh** [comment]
> * [x]  `hosted-rpi-arm64` (renamed from `hosted-rpi`)

What is the purpose of this "rpi" target? I don't see any difference, except that it includes the wiringpi GPIO module, which is deprected since 2019 and also broken on Debian bookworm-based Rpi Linux distibutions.

I would suggest simply removing the `hosted-rpi`/`hosted-rpi-arm64` target (and maybe even the whole [wiringpi module](https://github.com/modm-io/modm/tree/7318c2885ef60c4dee99212e8bba5c8b1e6d70a5/src/modm/platform/gpio/rpi)).

**Comment by salkinium** [comment]
> because the x86_64 context switching (fiber/context_x86_64.cpp) is choosen for both the hosted-linux and hosted-linux-arm64 targets:

Because I'm stupid and I copied the x86_64 file *without* modifications. 🙈🙈🙈 OMG, I'm getting old… 👴

> I would suggest simply removing the `hosted-rpi/hosted-rpi-arm64` target.

Yeah, that make sense. We only had GPIO support anyways, so with that gone, there's no difference to `hosted-linux-arm64`. I'll remove the code.

**Comment by salkinium** [comment]
Could you try it again now? @rleh

**Comment by rleh** [comment]
Now it works on a Aarch64 linux system (_Raspberry Pi 4 8GB_ running _Raspbian OS 64-bit_)!

image

@salkinium Could you check if it still works on ARM-based MacOS? 😆

**Comment by salkinium** [comment]
Will do. I also read that GitHub has ARM64 macOS runners, I'll check that out tonight.

File: src/modm/processing/fiber/module.lb [hunk 1]
```diff
@@ -64,6 +63,12 @@ def build(env):
         env.substitutions["default_stack_size"] = 2**20 # 1MB
         env.template("context_x86_64.cpp.in")
 
+    elif "arm64" in core:
+        env.substitutions["stack_minimum"] = (20 + 2) * 8
+        env.substitutions["default_stack_size"] = 2**20 # 1MB
+        env.copy("context_arm64.cpp")
+        env.copy("context_arm64_asm.S")
+
```
Comment (rleh) [inline src/modm/processing/fiber/module.lb:73]:
```suggestion
    else:
        raise NotImplementedError("Unknown architecture, no fiber context switch implementation available")
```

**Review by rleh — APPROVED** [review]
Thank you a lot for the work @salkinium!

**Comment by salkinium** [comment]
Ok, so.

GCC on macOS ignores `__attribute__((naked))`. Not sure why. That's why I moved the `modm_context_entry` function into assembly. I now found out that I can add asm statements into the `.cpp` file too, so I got rid of the `.S` file. I hope this also works on Linux GCC.

The macOS linker seems to require leading underscores. 🙄  This is also done in the boost asm files. I've added a template switch.

Could you test again on Linux?

**Comment by rleh** [comment]
> Could you test again on Linux?

Yes, and it works fine!

---

> The macOS linker seems to require leading underscores. 🙄 This is also done in the boost asm files. I've added a template switch.

GCC has a switch [`-fleading-underscore`](https://gcc.gnu.org/onlinedocs/gcc/Code-Gen-Options.html#index-fleading-underscore) to control the prefix of the emitted objects, but I would rather not touch it.

Because there is a [better option](https://gcc.gnu.org/onlinedocs/gcc/Asm-Labels.html):
```c++
extern "C" void modm_context_entry() asm("_modm_context_entry");
```

I reverted the template file back to a normal C++ file and implemented the ASM labels instead (see 1e3643455),  works (still) on arm64 linux.
Could you test again on MacOS? (Otherwise just drop my commit.)

**Comment by salkinium** [comment]
Yup, that works! Thanks! I'll squash and merge it after the CI passes.

**Comment by salkinium** [comment]
Removing the underscores also works this way!

**Review by rleh — APPROVED** [review]

**Review by chris-durand — APPROVED** [review]
Nice, thanks!

**Pull request merged**